recipes/lora_finetune_distributed.py

```diff
@@ -894,6 +899,77 @@ def train(self) -> None:
 
         self._profiler.stop()
 
+    def _loss_step(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
+        # Shape [b, s], needed for the loss not the model
+        labels = batch.pop("labels")
+
+        with self.activations_handling_ctx:
+            logits = self._model(**batch)
+
+        # Shift labels to compute loss
+        # equivalent to doing labels[..., 1:] and logits[..., :-1, :]
+        # But this way we dont need to slice the logits. We just add an ignore index to labels.
+        labels = torch.hstack(
+            (labels[..., 1:], self.ignore_labels_cache[: labels.shape[0]])
+        )
+        if not isinstance(logits, list):
+            labels = labels.reshape(-1)
+            logits = logits.reshape(-1, logits.size(-1))
+
+        # Compute loss
+        loss = self._loss_fn(logits, labels)
+        # free logits otherwise it peaks backward memory
+        del logits
+
+        return loss
+
+    def validate(self) -> Dict[str, float]:
+        """
+        Run validation loop and return average validation loss.
+        """
+        if self._val_dataloader is None:
```

Comment:
I'd rather throw an error upfront if the user tries to specify a "run_val_every_n_steps" without specifying a valdiation_dataset. This block of code should never be hit unless they're definitely doing validation.

Reply:
I found myself setting `run_val_every_n_steps` in config but then overriding `valdiation_dataset` to null from CLI (as a forgetful user who decided that's a way to disable validation temporarily).
But failing fast with an error message sounds alright (although I wish those errors were more readable in distributed cases).

Addressed in [6b15365](https://github.com/pytorch/torchtune/pull/2464/commits/6b15365413187e9065ad8fe7694d7fe7ffdbfe3a)